First pass. Go easy on me.

Doc-Saintly opened the pull request:
Still lots of work to do, but to add more modules it's basically just:
* Add a view
* Add a factory method
* Add it to the resolver

Commit:
First pass. Go easy on me.

MikeJeffers commented on src/Perpetuum/Modules/SlotFlags.cs:
```diff
@@ -1,20 +1,24 @@
+
+namespace Perpetuum.Modules
 {
+    /// Doc - These are powers of 8
```
powers of 2?

Doc-Saintly replied:
Yep, I was drinking something that night apparently.

MikeJeffers commented on src/Perpetuum/EntityFramework/EntityFactory.cs:
```diff
@@ -66,6 +66,43 @@ public Entity Create(EntityDefault entityDefault,IIDGenerator<long> idGenerator)
-    }
-}
+        public Entity CreateForData(string definitionName, IIDGenerator<long> idGenerator) {
+            bool forData = true;
```
`fordata` check seems unnecessary as this whole function is for that purpose right?

Doc-Saintly replied:
Correct. I can research to see if manually unpacking the robot and initializing it will have the same results - but I was worried that any additional "touching" of the entity would make the game server realize it's not actually attached to anyone and throw errors. I'll research.

Doc-Saintly replied:
Removed this and replaced with Robot.Unpack.

MikeJeffers commented on src/Perpetuum/EntityFramework/EntityFactory.cs:
```diff
@@ -66,6 +66,43 @@ public Entity Create(EntityDefault entityDefault,IIDGenerator<long> idGenerator)
         }
-    }
-}
+        public Entity CreateForData(string definitionName, IIDGenerator<long> idGenerator) {
```
Probably worth a clear docstring about how this is called for and only for the data-dumper...

That said... is there a way to write this as an extension method that lives in the data-dumper project?  That way this won't be accidentally called from any core server methods

Doc-Saintly replied:
removed

MikeJeffers commented on Perpetuum.DataDumper/Views/ModuleERPDataView.cs:
```diff
@@ -0,0 +1,19 @@
+namespace Perpetuum.DataDumper.Views {
+    public class ModuleERPDataView : ActiveModuleDataView {
+
+        public double module_recovery_chemical { get; set; }
```
I'll assume the casing here is because the variable name is used for something?
Otherwise stick to WhateverThisIsCalled

Doc-Saintly replied:
I did it this way because it's what the examples on the mediawiki templates were using, and the casing here is reflected and used for the variable names. It seems common to use this in mediaWIki templates, but there's no reason we can't do something different.

MikeJeffers commented on Perpetuum.DataDumper/Factory/DataDumper.Resolver.cs:
```diff
@@ -0,0 +1,94 @@
+                int itemId = (int)categoryItem["definition"];
+
+                try {
+                    object currentObject = entityFactory.CreateForData(itemName, EntityIDGenerator.Random);
```
Ok, this whole part with the CreateForData and the type checking, I think, could handle typing better where you wouldn't need type-check like is.

Also there is a better idiom I think for checking types *and* casting using *is* or *as* with a variable declaration in line.. its a bit fancy but here are the docs on it: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast

MikeJeffers commented on Perpetuum.DataDumper/Factory/DataDumper.ModuleRemoteArmorRepair.cs:
```diff
@@ -0,0 +1,27 @@
+
+            newView.module_optimal_range = module.GetBasePropertyModifier(AggregateField.optimal_range).Value * 10;
+            newView.module_repair_amount = module.GetBasePropertyModifier(AggregateField.armor_repair_amount).Value;
+            // module.GetBasePropertyModifier(AggregateField.heal)
```
?

Doc-Saintly replied:
I guess the question is about the * 10? For some reason all of the "range" are stored in meters / 10. This could be refactored to a function like `RangePropertyToMeters(double value)` that just multiplies by 10 but makes it more clear what's going on.

MikeJeffers replied:
oh no that's fine, just the commented out heal is what I was referring to.  I did this review backwards apparently (see comment below on the same comment line)

MikeJeffers commented on Perpetuum.DataDumper/Factory/DataDumper.ModuleERP.cs:
```diff
@@ -0,0 +1,29 @@
+            newView.module_recovery_kinetic = module.GetBasePropertyModifier(AggregateField.kinetic_damage_to_core_modifier).Value * 100;
+            newView.module_recovery_seismic = module.GetBasePropertyModifier(AggregateField.explosive_damage_to_core_modifier).Value * 100;
+            newView.module_recovery_thermal = module.GetBasePropertyModifier(AggregateField.thermal_damage_to_core_modifier).Value * 100;
+            // module.GetBasePropertyModifier(AggregateField.heal)
```
If this is "health" I think that is the repair state of the module/item.  Which for the data dump is likely irrelevant

Doc-Saintly replied:
You're talking about the `.heal` attribute? It's just left over junk when I was looking for the properties. Basically each time I add a new item to dump I add the view, the standard factory, and then set a breakpoint to look around at how to grab what I need.

MikeJeffers replied:
oh gotcha, debugging code.  Should clean some of that up to not mislead others to think its something important.

MikeJeffers commented on Perpetuum.DataDumper/Factory/DataDumper.AmmoWeapon.cs:
```diff
@@ -0,0 +1,53 @@
+            if (newView.damage_thermal == 0) { newView.damage_thermal = null; };
+            if (newView.damage_toxic == 0) { newView.damage_toxic = null; };
+
+            if (newView.modifier_falloff == "0") { newView.modifier_falloff = null; }; // keeping this as a string for now because other modules may use it as a modifier?
```
tricky....

MikeJeffers reviewed (changes requested):
Just a first pass for now, statically reading, I'll pull the code soon (tm)
For now just want to hear your thoughts on some of your decisions and if parts could get a little tighter
The big thing that sticks out is that new method on the entityframework and the way its used typechecking a raw object return.

MikeJeffers commented on Perpetuum.DataDumper/DataDumper.cs:
```diff
@@ -0,0 +1,604 @@
+
+            itemNames = new Dictionary<string, string>();
+
+            var dataLines = System.IO.File.ReadAllLines(@"C:\Users\derek\Desktop\projects\Perpetuum\dictionary.tsv");
```
I'll let some stuff slide for this first pass, but this should be revised ;)

MikeJeffers replied:
Actually I see what this is now, and maybe this is a bit trickier because its effectively an arbitrary file that could exist anywhere, but is needed for this step.
Probably allow the user to plug in the path, and/or check for presence in some default relative path and alert the user the file isnt found.

MikeJeffers commented on Perpetuum.DataDumper/DataDumper.cs:
```diff
@@ -0,0 +1,604 @@
+            }
+
+            // Now read the names from JSON files
+            var jsonFiles = System.IO.Directory.GetFiles(@"C:\Users\derek\Desktop\projects\Perpetuum\OPDB\Patches", "0.json", System.IO.SearchOption.AllDirectories);
```
This one is tricky to pull from another repo, you should use the 0.json as-installed on the perp server folder:
`/path/to/perpetuumserver/data/customDictionary/0.json`
The server reads it from this path in the data folder, so you can too!

MikeJeffers reviewed (changes requested):
Finally pulled the code, some small fixes are needed.

Commit:
Adding user settings manage paths

Changed localization parser to use GenXy deserializer
Added driller module
Removed changes to entity factory
Fixed SlotFlags comment
Removed some unnecessary comments

Commit:
Refactored. Robots don't export but it needs to be re-written anyway. Still need to update property names.

Commit:
Started to refactor properties to Camel Casing

Added option to open exported file
Added more module categories

DerekJarvis committed:
Fixed Demobilizer stat display

DerekJarvis committed:
Updated Data Dumper to work with robots

DerekJarvis committed:
Robot data updates

DerekJarvis committed:
Merge commit '8838baaa644b60ae3e89bfb18fca951dfa66b31f' into data-dumper

MikeJeffers commented on Perpetuum.DataDumper/Views/RobotDataView.cs:
```diff
@@ -0,0 +1,142 @@
+
+            //  Same as input.GetPropertyModifier(AggregateField.locking_range).Value * 10;
+            LockingRange = input.MaxTargetingRange * 10; 
+            LockingTime = input.GetPropertyModifier(AggregateField.locking_time).Value / 1000;
```
The `AggregateField` should have things like display units or display multiplier, which are transforms on the number for the purposes of rendering in a view for users.

MikeJeffers reviewed (commented):
Just a note on all the hardcoded coefficients on the data for display are stored in the DB.
And it looks like the client might actually apply the transforms as the only server side usage appears here: https://github.com/OpenPerpetuum/PerpetuumServer/blob/Development/src/Perpetuum.RequestHandlers/GetAggregateFields.cs#L24-L28
Might save time on transforming other data/fields later.

DerekJarvis committed:
Weapons view updated to new format. Updated on Wiki.
